.changeset/little-hornets-give.md

```diff
@@ -0,0 +1,34 @@
+---
+"astro": minor
+---
+
+Astro now allows to create your know i18n middleware:
+
+```js
+import { defineConfig } from "astro/config"
+// astro.config.mjs
+export default defineConfig({
+    i18n: {
+        locales: ["en", "fr"],
+        defaultLocale: "fr",
+        routing: "manual"
+    }
+})
+```
+
+```js
+// middleware.js
+import { redirectToDefaultLocale } from "astro:i18n";
+export const onRequest = defineMiddleware(async (context, next) => {
+    if (context.url.startsWith("/about")) {
+        return next()
+    } else {
+        return redirectToDefaultLocale(context, 302);  
+    }
+})
+```
+
+When `routing: "manual"` is provided, the virtual module `astro:i18n` exports new functions:
+- `redirectToDefaultLocale`
+- `notFound`
+- `redirectToFallback`
```

Comment:
``````suggestion
Also adds a `middleware` function that manually creates Astro's i18n middleware. This allows you to extend Astro's i18n routing instead of completely replacing it. Run `middleware` in combination with your own middleware, using the `sequence` utility to determine the order:

```js title="src/middleware.js"
import {defineMiddleware, sequence} from "astro:middleware";
import { middleware } from "astro:i18n"; // Astro's own i18n routing config

export const userMiddleware = defineMiddleware();

export const onRequest = sequence(
  userMiddleware,
  middleware({
    redirectToDefaultLocale: false,
    prefixDefaultLocale: true
  })
)
```
``````
This is a super stripped down version of the example in docs, but I think is worth showing here too as it's a feature of this PR!